sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -465,13 +465,18 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
       |   $initRangeFuncName(partitionIndex);
       | }
       |
-      | while (true) {
+      | while (true && !stopEarly()) {
       |   long $range = $batchEnd - $number;
       |   if ($range != 0L) {
       |     int $localEnd = (int)($range / ${step}L);
       |     for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
       |       long $value = ((long)$localIdx * ${step}L) + $number;
+      |       $numOutput.add(1);
```

Comment:
I have no worry about it since it is a simple op.

Reply:
This is very likely to hit perf regression since it's not a tight loop anymore.

We want the range operator to stop earlier for better performance, but it doesn't mean the range operator must return exactly the `limit` number of records. Since the range operator is already returning data in batch, I think we can stop earlier in a batch granularity.

Reply:
ok. then I should revert the `numOutput` change if the number of records can be a bit inaccurate.